spec/index.bs

```diff
@@ -11985,6 +12002,132 @@ It must only be included by interfaces which also include those mixins.
                     with the states from |passState| and |renderState|.
             </div>
         </div>
+
+    : <dfn>multiDrawIndirect(indirectBuffer, indirectOffset, maxDrawCount, drawCountBuffer, drawCountOffset)</dfn>
+    ::
+        Draws primitives using an array of parameters read from a {{GPUBuffer}}.
+        See [[#rendering-operations]] for the detailed specification.
+
+        This method is defined if and only if the {{GPUFeatureName/"multi-draw-indirect"}} [=feature=] is enabled.
+
+        The <dfn dfn for=>indirect multiDraw parameters</dfn> encoded in the |indirectBuffer| must be zero or more
+        tightly packed blocks of **four 32-bit unsigned integer values (16 bytes total)**, given in the same
+        order as the arguments for {{GPURenderCommandsMixin/draw()}}. For example:
+
+        <pre highlight="js">
+            let drawIndirectParameters = new Uint32Array(8);
+            // first draw
+            drawIndirectParameters[0] = vertexCount0;
+            drawIndirectParameters[1] = instanceCount0;
+            drawIndirectParameters[2] = firstVertex0;
+            drawIndirectParameters[3] = firstInstance0;
+            // second draw
+            drawIndirectParameters[4] = vertexCount1;
+            drawIndirectParameters[5] = instanceCount1;
+            drawIndirectParameters[6] = firstVertex1;
+            drawIndirectParameters[7] = firstInstance1;
+        </pre>
+
+        The <dfn dfn for=>indirect multiDraw count</dfn> encoded in the |drawCountBuffer| must be a **32-bit unsigned
+        integer** giving the number of draws to issue from the |indirectBuffer|.
+
+        <div algorithm="GPURenderCommandsMixin.multiDrawIndirect">
+            **Called on:** {{GPURenderCommandsMixin}} this.
+
+            **Arguments:**
+            <pre class=argumentdef for="GPURenderCommandsMixin/multiDrawIndirect(indirectBuffer, indirectOffset, maxDrawCount, drawCountBuffer, drawCountOffset)">
+                |indirectBuffer|: Buffer containing the [=indirect multiDraw parameters=].
+                |indirectOffset|: Offset in bytes into |indirectBuffer| where the drawing data begins.
+                |maxDrawCount|: Maximum number of draws or exact number of draws if |drawCountBuffer| is `null`.
+                |drawCountBuffer|: Buffer containing the [=indirect multiDraw count=].
+                |drawCountOffset|: Offset in bytes into |drawCountBuffer| where the [=indirect multiDraw count=] is located.
+            </pre>
+
+            **Returns:** {{undefined}}
+
+            Issue the following steps on the [=Device timeline=] of |this|.{{GPUObjectBase/[[device]]}}:
+            <div class=device-timeline>
+                1. If any of the following conditions are unsatisfied, make |this| [=invalid=] and stop.
+                    <div class=validusage>
+                        - It is [$valid to draw$] with |this|.
+                        - |indirectBuffer| is [$valid to use with$] |this|.
```

Comment:
(Opening a thread here because the threads on this issue still track all the things we need to do before landing this feature in the spec.)

According to https://github.com/gpuweb/gpuweb/issues/1354#issuecomment-2480937048 there may be limitations on what can be done on a command buffer in Metal if `supportIndirectCommandBuffers` is set to true. If so we may need to account for them in the spec. When we get a chance we'll look into the issue/repro on the [Chromium issue](https://crbug.com/379656378) and see if we think it needs to be accounted for in the spec. (Or maybe someone can find out whether wgpu has this problem and has already found a limitation or workaround.)

Reply:
@kainino0x WebKit always sets `supportIndirectCommandBuffers` to `YES` for all PSOs and we have no such issue problem running three.js's examples - https://github.com/search?q=repo%3AWebKit%2FWebKit%20supportIndirectCommandBuffers&type=code

I think it is because our implementation is based on Argument Buffers, if Chrome is planning to adopt that then that would be an option without changing the specification.

But setting `supportIndirectCommandBuffers` on the PSO does have a performance penalty, so I wouldn't rule out a spec change if we don't want to always enable this.

Reply:
That's super useful info, thank you!

Reply:
One such option would be extending [GPUShaderModuleCompilationHint](https://www.w3.org/TR/webgpu/#shader-module-compilation-hints) to indicate whether or not an entry point from a given shader module will be used with MTLIndirectCommandBuffer

If we want to explore that route, it would be nice to apply this to GPURenderBundle as well. I.e., if you set the flag to `0` indicating you will not use ICBs with a given PSO then you can not call `executeBundles(...)` from the `RenderPassEncoder` which uses that given entry point.

Then at GPURenderPipelineCreation, an implementation can lookup if the given entry point needs to set `supportIndirectCommandBuffers` to `YES` or not and potentially impact codegen as well